workflow/controller/operator.go

```diff
@@ -1739,6 +1739,11 @@ func getTemplateOutputsFromScope(tmpl *wfv1.Template, scope *wfScope) (*wfv1.Out
 		for _, art := range tmpl.Outputs.Artifacts {
 			resolvedArt, err := scope.resolveArtifact(art.From)
 			if err != nil {
+				// If the artifact was not found and is optional, don't mark an error
+				if strings.Contains(err.Error(), "Unable to resolve") && art.Optional {
```

Comment:
Even though we are doing string comparison to distinguish the error here, this behavior is covered by a test. After _much_ though and considering other alternatives (such as creating a custom error type), I decided this is the best way to go forward. Opinions welcome.